compiler/luci/pass/src/SubstituteTransposeToReshapePass.cpp

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2021 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/SubstituteTransposeToReshapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+
+namespace
+{
+
+/**
+ * @brief Convert transpose op in a certain condition to reshape op
+ * @details Convert transpose op if it have condition below
+ *          1. have a CircleConst perm value.
+ *          2. input have a unknown shape less then 2
+ *          3. the order of shape that except dim value 1 remains same on input and output
+ *          eg) input shape  = (1, 2, 3, 4, 5, 1, 1, 1) => (2, 3, 4, 5)
+ *              output shape = (2, 1, 3, 1, 1, 4, 5, 1) => (2, 3, 4, 5)
+ */
+bool substitute_transpose_to_reshape(luci::CircleNode *node)
+{
+  auto target_node = dynamic_cast<luci::CircleTranspose *>(node);
+  if (target_node == nullptr)
+    return false;
+
+  auto perm_const = dynamic_cast<luci::CircleConst *>(target_node->perm());
+  if (perm_const == nullptr)
+    return false;
+
+  assert(perm_const->dtype() == loco::DataType::S32);
+
+  auto input_node = loco::must_cast<luci::CircleNode *>(target_node->a());
+  if (perm_const->dim(0).value() != input_node->rank())
+    return false;
+  // If input have more than 2 unknown dimension, transpose will not changed.
+  int count = 0;
+  for (uint32_t i = 0; i < input_node->rank(); i++)
+    if (!input_node->dim(i).known())
+      count++;
+  if (count > 1)
+    return false;
+
+  uint32_t idx = 0;
+  for (uint32_t i = 0; i < perm_const->size<loco::DataType::S32>(); i++)
+  {
+    assert(perm_const->at<loco::DataType::S32>(i) >= 0 &&
+           perm_const->at<loco::DataType::S32>(i) < static_cast<int32_t>(input_node->rank()));
+    if (input_node->dim(static_cast<uint32_t>(perm_const->at<loco::DataType::S32>(i))).known() &&
+        input_node->dim(static_cast<uint32_t>(perm_const->at<loco::DataType::S32>(i))).value() == 1)
+      continue;
+    if (idx > static_cast<uint32_t>(perm_const->at<loco::DataType::S32>(i)))
```

Comment:
Maybe add `// To check idx values are increasing` ?

Reply:
That's right. I will add that comment.